x-pack/plugins/fleet/server/services/package_policies/populate_package_policy__assigned_agents_count.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core/server';
+
+import pMap from 'p-map';
+
+import type { PackagePolicy } from '../../../common';
+
+import { AGENTS_PREFIX } from '../../../common';
+import { getAgentsByKuery } from '../agents';
+
+/**
+ * Mutates each of the Package Policies passed on input and adds `agents` property to it with the
+ * count of agents currently using the given agent policy.
+ * @param esClient
+ * @param packagePolicies
+ */
+export const populatePackagePolicyAssignedAgentsCount = async (
+  esClient: ElasticsearchClient,
+  packagePolicies: PackagePolicy[]
+) => {
+  const agentPolicyRequestsByIdCache: Record<string, ReturnType<typeof getAgentsByKuery>> = {};
+
+  await pMap(
+    packagePolicies,
+    async (packagePolicy) => {
+      const agentPolicyId = packagePolicy.policy_id;
+      let agentsByQueryPromise: ReturnType<typeof getAgentsByKuery>;
+
+      if (Boolean(agentPolicyRequestsByIdCache[agentPolicyId])) {
+        agentsByQueryPromise = agentPolicyRequestsByIdCache[agentPolicyId];
+      } else {
+        agentsByQueryPromise = getAgentsByKuery(esClient, {
+          showInactive: false,
+          perPage: 0,
+          page: 1,
+          kuery: `${AGENTS_PREFIX}.policy_id:${agentPolicyId}`,
+        });
```

Comment:
Can we not use a terms aggregation to calculate this for all of the package policies in a a filter on a single search query?

Reply:
@joshdover,
Thanks for looking. Yeah, we probably can. Are you ok if we merge this as is now and I'll look at that approach next and push an update if I find a solution for it? 

_(I did cheat a bit on this one, as I just copied most of the code from the Agent Policies handler)_

Reply:
I know that we have customers with >1000 policies so this is likely to be an issue for them, but if we fix in an immediate follow up that'd be ok

Reply:
Ok 👍 
Let me look into this then before merging

Reply:
Hi @joshdover ,
Take another look when you get a chance. I added a new service function that uses aggregation to get the totals. let me know if that looks good to you.

Reply:
Looks great, thank you!